http://dev.mutt.org/trac/ticket/3314
http://dev.mutt.org/trac/changeset/053ef7bbaa72

Bail on missing pattern modifiers at end of pattern. Closes #3314.

Index: pattern.c
===================================================================
--- pattern.c (revision 5989:77ac8b5c2be6)
+++ pattern.c (revision 6018:053ef7bbaa72)
@@ -824,5 +824,11 @@
       case '=':
       case '~':
-	if (*(ps.dptr + 1) == '(') 
+	if (!*(ps.dptr + 1))
+	{
+	  snprintf (err->data, err->dsize, _("missing pattern: %s"), ps.dptr);
+	  mutt_pattern_free (&curlist);
+	  return NULL;
+	}
+	if (*(ps.dptr + 1) == '(')
         {
 	  ps.dptr ++; /* skip ~ */
